### PR TITLE
DO NOT MERGE - QUESTION: GI GETVECTOR machine scoring operator

### PR DESCRIPTION
**Pull request by agorina**

### CODE REVIEW

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        //crosses
+        assertEquals(geoObject.intersectsRegion(new Point(144, 338), new Point(300, 120), 5.0), true);
+        //contains
+        assertEquals(geoObject.intersectsRegion(new Point(149, 333), new Point(307, 199), 0.0), false);
+        //contains within 5 points
+        assertEquals(geoObject.intersectsRegion(new Point(149, 333), new Point(307, 199), 5.0), true);
+    }
+
+    @Test
+    public void testGetVector() throws TinyGRException {
```
Comment (agorina) [inline qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: While working on this operator in the new qti scoring engine, I wanted to test that the outcome is the same as the existing scoring engine, I am afraid I ran into a challenge.

The operator description states that:

> 

> Returns the index-th vector of a multi-vector object. Vectors are sorted starting with the vector
containing the top-most, left-most point, and **moving clockwise**. Where more than two line
segments meet at the same point, the top-most, left-most comes first.

We have also discussed the that current implementation supports closed 3-5 sides polygons. The below tests demonstrate that the **clockwise** rule in this scoring engine is implemented inconsistently. In fact, only one response in the below example is **clockwise*.

Comment (malaffoon) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: Nice ascii art!  There is a third option:
3. Assume the implementation is correct and try to understand how it jives with the specification. But i haven't looked at the examples/code yet.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            |__________________________________________________
+             |     |      |      |       |
+             20   30     45      60      90
+         **/
+
+        final String jai = "<Object><PointVector></PointVector>" +
+            "<EdgeVector>{ {(30,10),(20,50)} {(20,10),(20,50)} {(30,10),(20,10)}}</EdgeVector>" +
+            "</Object>";
+
+        // response is counter clock wise
```
Comment (malaffoon) [inline qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: Bit confused here. Is this a test of the existing scoring engine? Does this comment mean that the existing implementation returns `aij`?

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        assertEquals(TinyGR.getVector(jai, 0), "<Vector>{(20,50),(20,10)}</Vector>"); // reversed a
+        assertEquals(TinyGR.getVector(jai, 1), "<Vector>{(20,10),(30,10)}</Vector>"); // i
+        assertEquals(TinyGR.getVector(jai, 2), "<Vector>{(20,50),(30,10)}</Vector>"); // reversed j - why ?
```
Comment (malaffoon) [inline qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: Is the direction of the vector important? Seems to me that reversing (or not) is fine, as long as the order of the vectors is correct. Of course, i don't really understand why the order of the vectors matters either - any reasonable scoring engine should be able to figure out if the correct three vectors are specified.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java (hunk 1 above)
Comment (millerts) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: In a perfect world my vote is option 1.  However, I do think it is probably something to bring to Smarter's attention with the added information about how it impacts existing items.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java (hunk 2 above)
Comment (agorina) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: This is testing the existing scoring engine. And yes, the existing implementation returns "a,i,j".
The new implementation is there https://github.com/SmarterApp/AP_QTI/pull/139/files

Do not think I have cleared your confusion :) Let me know if you have more questions.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java (hunk 3 above)
Comment (agorina) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: I agree that the direction is not important. My point here was to compare with the new implementation ( the direction is also switched in there). I wanted to confirm that it is OK.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java (hunk 2 above)
Comment (malaffoon) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: Thanks, that unconfused me.

File: qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java (hunk 1 above)
Comment (wtritch) [reply qtiscoringengine/src/test/java/tinygrscoringengine/GeoObjectTest.java]: I just dug into the code.  It's... strange:
1. It picks the top-most, left-most vector first. (via `sortTopLeft`)
  a. It picks the top-most, then left-most point in each vector.   In the example, C and E share the top-most, left-most point.
  b. If point is shared, it compares "left-extent" which is the sum of the x-coordinates.  The smaller is chosen.  In the example, C and E have the same "left-extent" (30 + 45)
  c. If all else fails, it picks the shorter segment. In the example, this is C
2. The algorithm starts with the top-most-left-most point in the "first" vector as the initial "pivot-point".
3. The algorithm picks the next pivot-point as the "other" endpoint in the segment.  In the example (45, 80)
4. The algorithm sorts the remaining vectors according to which is closest to the current pivot-point. In the example this is F
5. Go to 3. and repeat until all vectors are sorted

In short, while it's more likely that the vectors will be sorted CCW, there's no enforcement in the implementation.  Any time the CW top-most, left-most vector in a shape has a smaller "left-extent" or shorter length for equal "left-extents" the vectors will be sorted CW.

FYI @alexdean201 this feels like a bug rather than a feature.  Since there are only 5 GI items in the production list that reference this method I'd like to stick to spec and enforce CCW edge sorting.
